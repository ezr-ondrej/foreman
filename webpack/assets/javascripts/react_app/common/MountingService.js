import ReactDOM from 'react-dom';
import store from '../redux';
import componentRegistry from '../components/componentRegistry';

export { default as registerReducer } from '../redux/reducers/registerReducer';

const mountedNodes = [];

// In order to support turbolinks with react, need to
// unmount all root components before turbolinks do the unload
// TODO: remove it when migrating into (webpacker-react or react-rails)
document.addEventListener('page:before-unload', () => {
  let node = mountedNodes.shift();

  while (node) {
    ReactDOM.unmountComponentAtNode(node);
    node = mountedNodes.shift();
  }
});

export function mount(component, selector, data, flattenData = false) {
  const reactNode = document.querySelector(selector);

  if (reactNode) {
    ReactDOM.unmountComponentAtNode(reactNode);
    ReactDOM.render(
      componentRegistry.markup(component, {
        data,
        store,
        flattenData,
      }),
      reactNode
    );

    mountedNodes.push(reactNode);
  } else {
    // eslint-disable-next-line no-console
    console.warn(
      `Cannot find '${selector}' element for mounting the '${component}'`
    );
  }
}

export function unmount(selector) {
  const reactNode = document.querySelector(selector);
  let i;

  if (reactNode) {
    for (i = 0; i < mountedNodes.length - 1; i++) {
      if (mountedNodes[i] === reactNode) {
        ReactDOM.unmountComponentAtNode(reactNode);
        mountedNodes.splice(i, 1);
        break;
      }
    }
  } else {
    // eslint-disable-next-line no-console
    console.warn(`Cannot find '${selector}' element for react unmounting`);
  }
}
