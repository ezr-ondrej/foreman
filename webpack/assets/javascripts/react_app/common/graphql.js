import React from 'react';
// import ApolloClient from 'apollo-boost';
import { ApolloClient } from 'apollo-client';
import { BatchHttpLink } from 'apollo-link-batch-http';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloLink, concat } from 'apollo-link';
import { ApolloProvider, Query } from 'react-apollo';

const httpLink = new BatchHttpLink({ uri: '/api/graphql' });
const authMiddleware = new ApolloLink((operation, forward) => {
  operation.setContext({
    headers: {
      'X-CSRF-Token': document
        .querySelector('meta[name=csrf-token]')
        .getAttribute('content'),
    },
  });
  return forward(operation);
});

const client = new ApolloClient({
  link: concat(authMiddleware, httpLink),
  cache: new InMemoryCache(),
});

const graphQLWrapperFactory = () => WrappedComponent => props => (
  <ApolloProvider client={client}>
    <WrappedComponent {...props} />
  </ApolloProvider>
);

const withGraphQLData = (Component, query) => props => (
  <Query query={query}>
    {({ loading, error, data }) => {
      if (loading) return 'Loading...';
      if (error) return `Error! ${error.message}`;
      return <Component {...props} graphqlData={data} />;
    }}
  </Query>
);

export { client, graphQLWrapperFactory, withGraphQLData };
