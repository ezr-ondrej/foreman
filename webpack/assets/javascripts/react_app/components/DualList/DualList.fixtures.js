export const props = {
  id: '123',
  items: [
    { label: 'location_01', value: '1' },
    { label: 'location_02', value: '2' },
    { label: 'location_03', value: '3' },
    { label: 'location_04', value: '4' },
  ],
  selectedIDs: [4],
  disabled: false,
};
