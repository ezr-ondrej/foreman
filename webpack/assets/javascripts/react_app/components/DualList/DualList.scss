.dual-list.disabled {
  pointer-events: none;
  cursor: not-allowed;

  .dual-list-pf-item {
    background: #f5f5f5;
    color: #8b8d8f;
  }
}
