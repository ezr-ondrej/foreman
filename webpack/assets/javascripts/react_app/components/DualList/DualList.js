import React from 'react';
import PropTypes from 'prop-types';
import { DualListControlled as PfDualList } from 'patternfly-react';
import { arrangeItemsBySelectedIDs } from './DualListHelper';

import './DualList.scss';

const DualList = props => {
  const { id, items, selectedIDs, inputProps, disabled, error } = props;
  const { selectedList, unselectedlist } = arrangeItemsBySelectedIDs(
    items,
    selectedIDs
  );

  return (
    <div id={id} className={`dual-list ${disabled ? 'disabled' : ''}`}>
      <PfDualList
        allowHiddenInputs
        left={{
          items: unselectedlist,
          inputProps,
        }}
        right={{
          items: selectedList,
          inputProps,
        }}
      />
      <div className="dual_list_error">{error}</div>
    </div>
  );
};

DualList.propTypes = {
  id: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
    })
  ),
  selectedIDs: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  ),
  inputProps: PropTypes.object,
  disabled: PropTypes.bool,
  error: PropTypes.string,
};

DualList.defaultProps = {
  id: null,
  items: [],
  selectedIDs: [],
  inputProps: {},
  disabled: false,
  error: null,
};

export default DualList;
