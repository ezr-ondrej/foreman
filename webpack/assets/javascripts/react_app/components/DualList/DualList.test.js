import { testComponentSnapshotsWithFixtures } from '../../common/testHelpers';
import DualList from './DualList';
import { props } from './DualList.fixtures';

const fixtures = {
  'renders DualList': props,
};

describe('DualList', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(DualList, fixtures));
});
