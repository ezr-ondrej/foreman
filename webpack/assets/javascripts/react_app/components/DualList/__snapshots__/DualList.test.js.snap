// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DualList rendering renders DualList 1`] = `
<div
  className="dual-list "
  id="123"
>
  <DualListControlled
    allowHiddenInputs={true}
    left={
      Object {
        "inputProps": Object {},
        "items": Array [
          Object {
            "label": "location_01",
            "value": "1",
          },
          Object {
            "label": "location_02",
            "value": "2",
          },
          Object {
            "label": "location_03",
            "value": "3",
          },
          Object {
            "label": "location_04",
            "value": "4",
          },
        ],
      }
    }
    onChange={[Function]}
    onComponentInit={[Function]}
    onFilterChange={[Function]}
    onItemChange={[Function]}
    onMainCheckboxChange={[Function]}
    onSortClick={[Function]}
    right={
      Object {
        "inputProps": Object {},
        "items": Array [],
      }
    }
  />
  <div
    className="dual_list_error"
  />
</div>
`;
